### ISSUE TITLE
Better key derivation function?

### ISSUE THREAD

**Issue by cstork** [rovaughn/box#1]
Hi, from glancing at main.c it seems you're using crypto_hash_sha256 as the key derivation function in case of password entry.  Any reason you're not using scrypt/pbkdf2/bcrypt?  Thanks for sharing your code, Chris

**Comment by rovaughn** [comment]
This is because it isn't a password in an ordinary sense, since it's a secret key.  The hash is just to make sure the key is 32 bytes, as is required by crypto_box.

A key derivation function is useful if we wanted to check if the password was equal to a previously entered password, without storing the password in plain text.  This is a different use case, as we are not storing the password or key anywhere, we just need to convert the password to 32 bytes so it may be used as a key.

Even if we were storing the password on disk, and we hashed it with a key derivation function, someone who got a hold of the resulting hash would still be able to decrypt/encrypt messages, so it wouldn't improve security.

**Comment by cstork** [comment]
I was thinking of using the KDF to protect against direct brute force attacks where the attacker runs unbox over an encrypted file with sha256 digests of known passwords.  AFAIK, the attacker could do this very efficiently since sha256 and `crypto_secretbox_open` are designed to be fast.  However, KDFs are designed to curtail this attack by being computationally expensive.

**Comment by rovaughn** [comment]
I see; that's an interesting point.

If we used a KDF to hash the password, it would still need a salt, which would need to be the same across every invocation.  A random salt could be generated and placed in the source code; this way there'd be no existing database that could be used to help crack the file.  However, it'd still be possible to make a database of common passwords using that salt, if one wanted to crack a variety of boxed files.

It sounds reasonably more secure, so I'll consider implementing it.  My original policy was that the user would accept the responsibility of choosing a password of suitable entropy, but like I said, I'll think about it.

**Comment by cstork** [comment]
I agree that using a good salt would be a further improvement.  And I agree that using the same salt for all box-encrypted files has the trade-offs you describe. I see two more alternatives:
1. Let the user choose the salt explicitly. (Potentially confusing to non-experts.)  There could always be the fallback on the constant salt.
2. Do as the s/bcrypt file encryption utilities do and add a random salt to the header of your ciphertext.

(In any case, it would be best to advice the user to modify, or better, to augment their passwords with some "salt" that is input specific, e.g., the input file name.)

However, there could be a practical problem.  At least in the case of scrypt and bcrypt there don't seem to be libraries available for the actual KNFs. (That's my first impression after a non exhaustive search.) Both utilities do file encryption and use a similar scheme to encode a random salt in the header of the ciphertext.  See these two links for more on this:
- http://stackoverflow.com/questions/13654492/how-to-use-scrypt-to-generate-hash-for-password-and-salt-in-python
- http://stackoverflow.com/questions/6832445/how-can-bcrypt-have-built-in-salts

**Comment by adrianprv** [comment]
https://github.com/adrianprv/HKF-Memory-Hard-Password-Key-Derivation-Function
